source/transactions/transactions.rst

```diff
@@ -412,6 +412,15 @@ after commitTransaction fails with a retryable error according to the
 Retryable Writes Specification, regardless of whether retryWrites is set
 on the MongoClient or not.
 
+When commitTransaction is retried, either by the driver's internal retry-once
+logic or explicitly by the user calling commitTransaction again, drivers MUST
+apply ``w: majority`` to the write concern of the commitTransaction command. If
+the transaction is using a `writeConcern`_ that is not the server default (i.e.
+specified via TransactionOptions during the ``startTransaction`` call or
+otherwise inherited), any other write concern options (e.g. ``wtimeout``) MUST
+be left as-is when applying ``w: majority``. See
+`Majority write concern is used when retrying commitTransaction`_.
```

Comment:
A write concern tag set is expressed as a string via the `w` option, so it's mutually exclusive with a majority write concern. Furthermore, you actually can't have a custom tag set named "majority". I'm not sure if the server actually prevents you from defining one, but I do expect that a "majority" string for `w` will be interpreted as a "majority of the replica set" before consulting for a tag set with that name. See: https://docs.mongodb.com/manual/reference/write-concern/#w-option

Reply:
Thanks, but what about my other question? If the application is using `w:3` on a 3 node replica set and the driver replaces this with `w:majority` then the transaction will only wait for the write to be present on 2 nodes. Should we add a note explaining this subtlety?

Reply:
Happy to add a note about the edge case of altering a write concern to the Q&A section. Your question here overlaps with some discussion @p-mongo and I had in the Slack channel for the `withTransaction()` spec:

> Oleg: So for 1185, we are going to downgrade user's write concern to majority if they specified a number which is greater than majority?
>
> Jeremy: i don't recall if "majority" is special in this respect, beyond its normal specialness of always being the computed majority for the particular RS topology. I think this might have come up in since-resolved comments on the design document, but Tess could certainly confirm. That said, since a replica set could have hidden nodes (which afaik do count towards a majority), I don't think a driver could even reliably calculate the majority -- assuming Oleg is asking about having the driver possibly compare if the original write concern was > majority in order to preserve it
> 
>Jeremy: i'm also not sure if there's a case for a 5 node replica set (just for example) where w:4 has any benefit over w:majority. Possibly for geographic distribution but then I'd expect users to use a custom write concern string -- and with those a driver certainly doesn't know if such a string would be rollback-proof, which majority affords us. therefore, i think "majority" is always reasonable for our purpose of retrying

Aly pointed out that another special property of "majority" is that it implies journaling. That should probably be mentioned here as well. It also makes me wonder what would happen if the commit was originally using something like `{w: 2, j: false}` and we "upgraded" it to `{w: "majority", j: false}` -- would that result in an error, or would the server ignore `j: false`? Something I should test before finalizing this.

Reply:
Yeah this SGTM. Note that another option could be for drivers to issue a retry with `w:majority` and then perform a final attempt with the user's original write concern. This would ensure that both `w:majority` and the user's write concern are respected but I think this would be too much complication for too little benefit.

Reply:
I concur that the final attempt with the original write concern would invite more complexity than it's worth. What if _that_ fails? Do we end up retrying a `w:majority` for no tangible benefit?

I've also noted discussion in Slack about [`writeConcernMajorityJournalDefault)`](https://docs.mongodb.com/manual/reference/replica-configuration/#rsconf.writeConcernMajorityJournalDefault). While `w:majority` may default to `j:true`, it's still kosher to pass `{ w: "majority", j: false }` as a write concern. Since it's unlikely for users to customize `writeConcernMajorityJournalDefault`, I think we're fine relying on the existing design of only overriding `w` and leaving other WC options as-is.

Reply:
I added a final paragraph in "Majority write concern is used when retrying commitTransaction" to elaborate on why we're OK with applying "majority" over any existing `w` value.

Reply:
From Judah on slack (paraphrased a bit):
Me:
> We have a question that’s related to a drivers spec change: how does writeConcern:{w:“majority”} differ from say {w:2}? AFAIK w:majority is “special” in that it updates the primary’s majority commit point whereas a w:N write may not.
> More specifically, we are wondering whether the edge case described in https://jira.mongodb.org/browse/SPEC-1185 would still be possible if the transaction was committed with {w:2} instead of {w:1} (assuming this is a 3 node replica set).

Judah:
> w:majority waits on the majority commit point, rather than just what gets replicated to a majority of nodes. In general those are mostly the same. w:majority waits a little longer to get read your own writes for read concern majority, but that doesn’t give actual guarantees. nodes also wait for the “first optime of their term” to be committed for w:majority but not for w:2.
> So the short answer is, w:2 may be safe, but w:majority definitely is, without me looking a lot more into that Spec